### PR TITLE
[Config] Add some cache on SelfCheckingResourceChecker

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

https://blackfire.io/profiles/compare/74eb90ee-7367-45dd-ac63-7145353b1ffa/graph

### CODE REVIEW

File: src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php [hunk 1]
```diff
@@ -23,14 +23,30 @@
...
+    private static array $cache = [];
+
     public function supports(ResourceInterface $metadata)
     {
         return $metadata instanceof SelfCheckingResourceInterface;
     }
 
     public function isFresh(ResourceInterface $resource, int $timestamp)
     {
+        $key = "$resource:$timestamp";
```
Comment (javiereguiluz) [inline src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php:42]: What's the `$timestamp` resolution here? If two different checkers check the same resource but some nanoseconds later, would the key be different?

File: src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php [hunk 2]
```diff
@@ -23,14 +23,30 @@
...
+        if (\array_key_exists($key, self::$cache)) {
+            return self::$cache[$key];
+        }
+
         /* @var SelfCheckingResourceInterface $resource */
-        return $resource->isFresh($timestamp);
+        return self::$cache[$key] = $resource->isFresh($timestamp);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php]: no need for array_key_exists, this should do it, isn't it?
`return self::$cache[$key] ?? self::$cache[$key] = $resource->isFresh($timestamp);`

File: src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php [hunk 3]
```diff
@@ -23,14 +23,30 @@
...
+    }
+
+    public static function reset()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php]: not sure we want to expose this in the API
maybe use reflection in the tests instead?

File: src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php [hunk 4]
```diff
@@ -23,14 +23,30 @@
  */
 class SelfCheckingResourceChecker implements ResourceCheckerInterface
 {
+    // Common shared cache, because this checker can be used in different
+    // situation. For example, when using the full stack framework, the router
+    // and container have their own cache. But they may check on the very same
+    // resources
+    private static array $cache = [];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php]: hello PHP 7.* :)

File: src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php (hunk 1 above)
Comment (lyrixx) [reply src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php:42]: I checked, and on a regular Symfony application, the timestamp is the same. and the cache behaves as expected. So I guess it's fine

**Comment by lyrixx** [comment]
@nicolas-grekas Thanks for the review. I updated the PR

**Review by dunglas — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php [hunk 5]
```diff
@@ -23,14 +23,24 @@
  */
 class SelfCheckingResourceChecker implements ResourceCheckerInterface
 {
+    // Common shared cache, because this checker can be used in different
+    // situation. For example, when using the full stack framework, the router
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php]: situations (missing "s")

File: src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php [hunk 6]
```diff
@@ -23,14 +23,24 @@
...
+    // and container have their own cache. But they may check on the very same
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/SelfCheckingResourceChecker.php]: `and *the* container`

`s/ on / /`

**Review by nicolas-grekas — COMMENTED** [review]
not pretty :)
aren't there a way to create a shared SelfCheckingResourceChecker and make this property non-static?

**Comment by lyrixx** [comment]
> not pretty :)

Yes! I agree, I'm not really proud of it ! 

> aren't there a way to create a shared SelfCheckingResourceChecker and make this property non-static?

This class is instantiated in https://github.com/symfony/symfony/blob/12f8faefd0cf539c3cecbdc375f87605c96a4376/src/Symfony/Component/Config/ConfigCache.php#L40 when symfony is in debug mode

and `ConfigCache` is instantiated in many different location (config, kernel, and routing component).

So I guess we may find a way to share it, but the code will be even uglier

**Review by nicolas-grekas — APPROVED** [review]
Ok, go on my side, we'll always be able to improve later on if needed.

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Pull request merged**